src/back/space/src/service/service.cpp

```diff
@@ -0,0 +1,330 @@
+	return _canCreate;
+}
+
+int Service::CountInvitationAvailable(const std::string& currentUserId) {
+	return _repo.SpaceInvitation().GetAvailableCount(currentUserId);
+}
+
+bool Service::KeyCreateCheck(const std::string& key, const std::string& userId) {
+	// При создании ключ может либо совпадать с userId
+	if (key == userId)
+		return true;
+
+	// либо не должен быть зарезервирован
+	if (isKeyReserved(key))
+		return false;
+
+	// и содержать только допустимые символы
+	static const std::regex rx("[a-z0-9][a-z0-9_]+");
+	return std::regex_match(key, rx);
+}
+
+bool Service::KeyWeakCheck(const std::string& key) {
+	// При получении Space, ключ не должен быть зарезервирован
+	if (isKeyReserved(key))
+		return false;
+	// и содержать ослабленный вариант допустимых символов
+	static const std::regex rx("[a-z0-9_-]+");
+	return std::regex_match(key, rx);
+}
+
+bool Service::IsUserTimeouted(const std::string& userId) {
+	return _repo.IsReadyForCreationByTime(userId);
+}
+
+bool Service::IsLimitReached(const std::string& userId) {
+	return _spacesLimitForUser <= _repo.GetCountSpacesWithUser(userId);
+}
+
+void Service::Create(const std::string& name, const std::string& key, bool requestsAllowed, const std::string& userId) {
+	_repo.CreateSpaceAndItsOwner(name, key, requestsAllowed, userId);
+}
+
+void Service::Delete(const boost::uuids::uuid& id) {
+	_repo.SpaceUser().DeleteBySpace(id);
+	_repo.SpaceInvitation().DeleteBySpace(id);
+	_repo.SpaceLink().DeleteBySpace(id);
+	_repo.Space().Delete(id);
+}
+
+bool Service::IsSpaceOwner(const boost::uuids::uuid& id, const std::string& userId) {
+	return _repo.SpaceUser().IsOwner(id, userId);
+}
+
+void Service::Invite(const std::string& creatorId, const boost::uuids::uuid& spaceId, const std::string& userId, const Role::Type& role) {
+	_repo.SpaceInvitation().Insert(spaceId, userId, role, creatorId);
+}
+
+void Service::ChangeRoleInInvitation(int id, const Role::Type& role, const std::string& userId) {
+	const auto invitation = _repo.SpaceInvitation().SelectById(id);
+	if (!_repo.SpaceUser().IsAdmin(invitation.spaceId, userId))
+		throw errors::Unauthorized();
+
+	_repo.SpaceInvitation().UpdateRole(id, role);
+}
+
+void Service::ApproveInvitation(int id, const std::string& headerUserId) {
+	// todo - rewrite into 1 transaction
+	model::SpaceInvitation invitation = _repo.SpaceInvitation().SelectById(id);
+
+	bool isEnoughRights = false;
+
+
+	if (invitation.creatorId == invitation.userId && invitation.userId != headerUserId) {
+		if (_repo.SpaceUser().IsAdmin(invitation.spaceId, headerUserId))
+			isEnoughRights = true;
+	} else if (invitation.creatorId != invitation.userId && invitation.userId == headerUserId)
+			isEnoughRights = true;
+
+	if (!isEnoughRights)
+		throw errors::Unauthorized();
+
+	static const std::set<std::string> valid_roles{
+		"user", "guest", "admin"
+	};
+	if (!valid_roles.contains(Role::ToString(invitation.role)))
+		throw errors::BadRequest("Wrong role");
```

Comment:
Нужно стараться не использовать сравнение по строкам там где можем использовать преимущества строгой типизации.

Перенеси пожалуйста эту проверку перед проверкой прав. 
```suggestion
	static const std::set<Role::Type> valid_roles{
		Role::User, Role::Guest, Role::Admin
	};
	if (!valid_roles.contains(invitation.role))
		throw errors::BadRequest("Wrong role");
```

Reply:
Готово.